### PR TITLE
Use Context for Minikube clusters

### LINKED ISSUE #290: Don't have the option "Set as current cluster"
Build: 4651

1. Open the Kubernetes: cluster explorer
2. Right click the "minikube" cluster

Don't have the option "Set as current cluster"
image

But right click other cluster have the option "Set as current cluster"

### PR DESCRIPTION
**Pull request by bnookala**
Fixes #290 & adds the minikube logo.

### CODE REVIEW

File: src/explorer.ts [hunk 1]
```diff
@@ -32,7 +35,7 @@ export interface KubernetesObject {
     readonly id: string;
     readonly metadata?: any;
     getChildren(kubectl: Kubectl, host: Host): vscode.ProviderResult<KubernetesObject[]>;
-    getTreeItem(): vscode.TreeItem | Thenable<vscode.TreeItem>;
+    getTreeItem(type: string | undefined): vscode.TreeItem | Thenable<vscode.TreeItem>;
```
Comment (itowlson) [inline src/explorer.ts]: I'm not keen on this - the only place the `type` information can come from is surely the object itself, so why would it need to pass it to its own member function?  Suggest that things that are interested in `type` implement it as a field and inspect it from their `getTreeItem` implementations...?

File: src/explorer.ts [hunk 2]
```diff
@@ -117,23 +120,25 @@ class KubernetesContext implements KubernetesObject {
         ];
     }
 
-    getTreeItem(): vscode.TreeItem | Thenable<vscode.TreeItem> {
+    getTreeItem(type: string | undefined): vscode.TreeItem | Thenable<vscode.TreeItem> {
         const treeItem = new vscode.TreeItem(this.id, vscode.TreeItemCollapsibleState.Collapsed);
-        treeItem.contextValue = "vsKubernetes.cluster";
+        treeItem.contextValue = type ? type : KUBERNETES_CLUSTER;
```
Comment (itowlson) [inline src/explorer.ts]: This definitely feels like it should be based on a `this.type` field - the type is an attribute of the object, not a transient parameter that might vary from method call to method call.

**Comment by bnookala** [comment]
@itowlson good points all around. I've changed the implementation to align with your request.

File: src/explorer.ts [hunk 3]
```diff
@@ -70,12 +73,13 @@ export class KubernetesExplorer implements vscode.TreeDataProvider<KubernetesObj
 
     private async getClusters(): Promise<KubernetesObject[]> {
         const contexts = await kubectlUtils.getContexts(this.kubectl);
-        return contexts.map((context): KubernetesContext => {
+        return contexts.map((context): ClusterContext => {
```
Comment (itowlson) [inline src/explorer.ts]: This type declaration seems wrong (and seems like it was wrong before).  `kubectlUtils.getContexts` returns `kubectlUtils.KubectlContext[]` - a model object not an explorer object.  (But funnily enough it seems like TypeScript is ignoring the type declaration and inferring `context` as `KubectlContext`... weird...)

File: src/explorer.ts [hunk 4]
```diff
@@ -101,8 +105,9 @@ class DummyObject implements KubernetesObject {
     }
 }
 
-class KubernetesContext implements KubernetesObject {
-    constructor(readonly id: string, readonly metadata?: kubectlUtils.KubectlContext) {
+class ClusterContext implements KubernetesObject {
```
Comment (itowlson) [inline src/explorer.ts]: I don't get the value in renaming this.  If we're going to rename it, I'd suggest something like `ContextNode` or `ContextObject` or `KubernetesContextObject`.

File: src/explorer.ts [hunk 5]
```diff
@@ -119,21 +124,29 @@ class KubernetesContext implements KubernetesObject {
 
     getTreeItem(): vscode.TreeItem | Thenable<vscode.TreeItem> {
         const treeItem = new vscode.TreeItem(this.id, vscode.TreeItemCollapsibleState.Collapsed);
-        treeItem.contextValue = "vsKubernetes.cluster";
-        treeItem.iconPath = vscode.Uri.file(path.join(__dirname, "../../images/k8s-logo.png"));
+        treeItem.contextValue = this.clusterType;
+
+        let iconPath: vscode.Uri;
```
Comment (itowlson) [inline src/explorer.ts]: Consider breaking this block out into a separate function that returns the icon path, so we can make this `const` - a la `const iconPath = getIconPathForClusterType();`

File: src/explorer.ts (hunk 3 above)
Comment (bnookala) [reply src/explorer.ts]: It seems like it's inferring the correct type for me…

File: src/explorer.ts (hunk 5 above)
Comment (bnookala) [reply src/explorer.ts]: fixed.

File: src/explorer.ts (hunk 4 above)
Comment (bnookala) [reply src/explorer.ts]: fixed.

**Pull request merged**